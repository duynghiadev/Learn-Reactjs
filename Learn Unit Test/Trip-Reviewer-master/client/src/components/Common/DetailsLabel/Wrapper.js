import styled from 'styled-components';
import PropTypes from 'prop-types';

const Wrapper = styled.div`
  background: ${({ color }) => color};
  display: inline-block;
  padding: 5px 10px;
  border-radius: 3px;
  font-size: 12px;
  margin-right: 5px;
  color: #fff;
`;

Wrapper.propTypes = {
  color: PropTypes.string.isRequired,
};

export default Wrapper;
