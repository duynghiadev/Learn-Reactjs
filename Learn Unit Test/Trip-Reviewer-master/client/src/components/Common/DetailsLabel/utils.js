export const TYPES = {
  expences: 'EXPENCES',
  duration: 'DURATION',
  rating: 'RATING',
};

const colors = {
  veryGood: '#00B300',
  good: '#7FBB17',
  ok: '#FEC22D',
  bad: '#F47048',
  veryBad: '#E91E63',
};

const limitsExpences = [100, 500, 1000, 5000];
const limitsDuration = [2, 7, 14, 31];
const limitsRating = [1, 2, 3, 4];

const getColorsInversed = (value, limits) => {
  switch (true) {
    case value > limits[3]:
      return colors.veryBad;
    case value > limits[2]:
      return colors.bad;
    case value > limits[1]:
      return colors.ok;
    case value > limits[0]:
      return colors.good;
    default:
      return colors.veryGood;
  }
};

const getColorsDirect = (value, limits) => {
  switch (true) {
    case value > limits[3]:
      return colors.veryGood;
    case value > limits[2]:
      return colors.good;
    case value > limits[1]:
      return colors.ok;
    case value > limits[0]:
      return colors.bad;
    default:
      return colors.veryBad;
  }
};

const getFunctionByType = {
  [TYPES.expences]: value => getColorsInversed(value, limitsExpences),
  [TYPES.duration]: value => getColorsDirect(value, limitsDuration),
  [TYPES.rating]: value => getColorsDirect(value, limitsRating),
};

export const getColorsByTypeAndValue = (type, value) => getFunctionByType[type](value);
