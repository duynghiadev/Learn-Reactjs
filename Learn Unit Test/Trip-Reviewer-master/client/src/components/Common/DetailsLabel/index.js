import React from 'react';
import PropTypes from 'prop-types';

import Wrapper from './Wrapper';
import { getColorsByTypeAndValue } from './utils';

const DetailsLabel = ({ children, type, value }) => {
  const color = getColorsByTypeAndValue(type, value);

  return <Wrapper color={color}>{children}</Wrapper>;
};

DetailsLabel.propTypes = {
  children: PropTypes.node.isRequired,
  type: PropTypes.string.isRequired,
  value: PropTypes.number.isRequired,
};

export default DetailsLabel;
