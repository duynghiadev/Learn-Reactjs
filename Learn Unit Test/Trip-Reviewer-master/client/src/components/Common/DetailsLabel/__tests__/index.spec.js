import React from 'react';
import { create } from 'react-test-renderer';

import DetailsLabel from '..';
import { TYPES } from '../utils';

const mockComponent = props => <DetailsLabel {...props} />;

describe('<DetailsLabel />', () => {
  describe('Duration', () => {
    const props = {
      children: 'children',
      type: TYPES.duration,
    };

    it('should render very bad duration', () => {
      props.value = 1;
      const tree = create(mockComponent(props)).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render bad duration', () => {
      props.value = 3;
      const tree = create(mockComponent(props)).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render ok duration', () => {
      props.value = 8;
      const tree = create(mockComponent(props)).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render good duration', () => {
      props.value = 15;
      const tree = create(mockComponent(props)).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render very good duration', () => {
      props.value = 32;
      const tree = create(mockComponent(props)).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('Expences', () => {
    const props = {
      children: 'children',
      type: TYPES.expences,
    };

    it('should render very bad expences', () => {
      props.value = 99;
      const tree = create(mockComponent(props)).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render bad expences', () => {
      props.value = 101;
      const tree = create(mockComponent(props)).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render ok expences', () => {
      props.value = 501;
      const tree = create(mockComponent(props)).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render good expences', () => {
      props.value = 1001;
      const tree = create(mockComponent(props)).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render very good expences', () => {
      props.value = 5001;
      const tree = create(mockComponent(props)).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('Rating', () => {
    const props = {
      children: 'children',
      type: TYPES.rating,
    };

    it('should render very bad rating', () => {
      props.value = 1;
      const tree = create(mockComponent(props)).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render bad rating', () => {
      props.value = 3;
      const tree = create(mockComponent(props)).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render ok rating', () => {
      props.value = 8;
      const tree = create(mockComponent(props)).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render good rating', () => {
      props.value = 15;
      const tree = create(mockComponent(props)).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render very good rating', () => {
      props.value = 32;
      const tree = create(mockComponent(props)).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
