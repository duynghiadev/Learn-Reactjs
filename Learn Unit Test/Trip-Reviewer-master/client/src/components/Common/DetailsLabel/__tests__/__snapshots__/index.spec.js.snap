// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetailsLabel /> Duration should render bad duration 1`] = `
<div
  className="Wrapper-ThVjl cjMZrE"
  color="#F47048"
>
  children
</div>
`;

exports[`<DetailsLabel /> Duration should render good duration 1`] = `
<div
  className="Wrapper-ThVjl gLyMCc"
  color="#7FBB17"
>
  children
</div>
`;

exports[`<DetailsLabel /> Duration should render ok duration 1`] = `
<div
  className="Wrapper-ThVjl elJdDI"
  color="#FEC22D"
>
  children
</div>
`;

exports[`<DetailsLabel /> Duration should render very bad duration 1`] = `
<div
  className="Wrapper-ThVjl hLebDS"
  color="#E91E63"
>
  children
</div>
`;

exports[`<DetailsLabel /> Duration should render very good duration 1`] = `
<div
  className="Wrapper-ThVjl gUEEXL"
  color="#00B300"
>
  children
</div>
`;

exports[`<DetailsLabel /> Expences should render bad expences 1`] = `
<div
  className="Wrapper-ThVjl gLyMCc"
  color="#7FBB17"
>
  children
</div>
`;

exports[`<DetailsLabel /> Expences should render good expences 1`] = `
<div
  className="Wrapper-ThVjl cjMZrE"
  color="#F47048"
>
  children
</div>
`;

exports[`<DetailsLabel /> Expences should render ok expences 1`] = `
<div
  className="Wrapper-ThVjl elJdDI"
  color="#FEC22D"
>
  children
</div>
`;

exports[`<DetailsLabel /> Expences should render very bad expences 1`] = `
<div
  className="Wrapper-ThVjl gUEEXL"
  color="#00B300"
>
  children
</div>
`;

exports[`<DetailsLabel /> Expences should render very good expences 1`] = `
<div
  className="Wrapper-ThVjl hLebDS"
  color="#E91E63"
>
  children
</div>
`;

exports[`<DetailsLabel /> Rating should render bad rating 1`] = `
<div
  className="Wrapper-ThVjl elJdDI"
  color="#FEC22D"
>
  children
</div>
`;

exports[`<DetailsLabel /> Rating should render good rating 1`] = `
<div
  className="Wrapper-ThVjl gUEEXL"
  color="#00B300"
>
  children
</div>
`;

exports[`<DetailsLabel /> Rating should render ok rating 1`] = `
<div
  className="Wrapper-ThVjl gUEEXL"
  color="#00B300"
>
  children
</div>
`;

exports[`<DetailsLabel /> Rating should render very bad rating 1`] = `
<div
  className="Wrapper-ThVjl hLebDS"
  color="#E91E63"
>
  children
</div>
`;

exports[`<DetailsLabel /> Rating should render very good rating 1`] = `
<div
  className="Wrapper-ThVjl gUEEXL"
  color="#00B300"
>
  children
</div>
`;
